imgui/src/drag_drop.rs

```diff
@@ -0,0 +1,475 @@
+    /// `ACCEPT_NO_PREVIEW_TOOLTIP`, which will also prevent this tooltip from being shown.
+    ///
+    /// This function also takes a payload in the form of `T: Copy + 'static`. We use this bound to
+    /// ensure that we can safely send and receive the given type from C++ without worrying about
+    /// handling drops.
+    pub fn begin_payload<'ui, T: Copy + 'static>(
+        self,
+        ui: &Ui<'ui>,
+        payload: T,
+    ) -> Option<DragDropSourceToolTip<'ui>> {
+        unsafe {
+            let payload = TypedPayload::new(payload);
+            self.begin_payload_unchecked(
+                ui,
+                &payload as *const _ as *const ffi::c_void,
+                std::mem::size_of::<TypedPayload<T>>(),
+            )
+        }
+    }
+
+    /// Creates the source of a drag and returns a handle on the tooltip.
+    /// This handle can be immediately dropped without binding it, in which case a default empty
+    /// circle will be used for the "blank" tooltip as this item is being dragged around.
+    ///
+    /// Otherwise, use this tooltip to add data which will display as this item is dragged.
+    /// If `SOURCE_NO_PREVIEW_TOOLTIP` is enabled, however, no preview will be displayed
+    /// and this returned token does nothing. Additionally, a given target may use the flag
+    /// `ACCEPT_NO_PREVIEW_TOOLTIP`, which will also prevent this tooltip from being shown.
+    ///
+    /// This function also takes a payload of any `*const T`. Please avoid directly using it
+    /// if you can.
+    ///
+    /// ## Safety
+    /// This function itself will not cause a panic, but using it directly opts you into
+    /// managing the lifetime of the pointer provided yourself. Dear ImGui will execute a memcpy on
+    /// the data passed in with the size (in bytes) given, but this is, of course, just a copy,
+    /// so if you pass in an `&String`, for example, the underlying String data will not be cloned,
+    /// and could easily dangle if the `String` is dropped.
+    ///
+    /// Moreover, if `Condition::Always` is set (as it is by default), you will be copying in your data
+    /// every time this function is ran in your update loop, which if it involves an allocating and then
+    /// handing the allocation to ImGui, would result in a significant amount of data created.
+    ///
+    /// Overall, users should be very sure that this function is needed before they reach for it, and instead
+    /// should consider either [begin](Self::begin) or [begin_payload](Self::begin_payload).
+    pub unsafe fn begin_payload_unchecked<'ui>(
+        &self,
+        _ui: &Ui<'ui>,
+        ptr: *const ffi::c_void,
+        size: usize,
+    ) -> Option<DragDropSourceToolTip<'ui>> {
+        let should_begin = sys::igBeginDragDropSource(self.flags.bits() as i32);
+
+        if should_begin {
+            sys::igSetDragDropPayload(self.name.as_ptr(), ptr, size, self.cond as i32);
```

Comment:
So, i think imgui c++ tends to truncate usize to 32 bits (often signed). We should probably have an assert somewhere that the size is within that... (I assume this is one of the places where that happens, since we have an i32 size later on).

... I bet our existing code mishandles this too in lots of places...

Reply:
left this unimplemented for now, mostly because it's just very late -- will add a note in a second